File: navdeepghai/dashboard/g_dashboard/book_appointment/book_appointment.js
/* write your js controller here */

frappe.ready(function(){

    frappe.provide("service_technician");

    service_technician.ServiceTechnician = germbusters.dashboard.Dashboard.extend({
        get_row: function(idx, columns, val){
            let $row = $(`<tr addr='${val.customer_address}' booking='${val.name}'>
                    <td style="${columns[0].column_width}% !important;">
                        <span>
                            <b>${__("Customer Name: ")}</b>
                            ${val.full_name}
                        </span><br>
                        <span>
                            <b>${__("Booking ID: ")}</b>
                            ${val.name}
                        </span><br>
                        <span>
                            <b>${__("Task ID: ")}</b>
                            ${val.task}
                        </span><br>
                        <span>
                            <b>${__("Mobile: ")}</b>
                            ${val.mobile}
                        </span><br>
                        <span>
                            <b>${__("Email: ")}</b>
                            ${val.email}
                        </span><br>
                        <span>
                            <b>${__("Total Targets: ")}</b>
                                ${val.total_targets || ""}
                        </span><br>
                        <span>
                            <b>${__("Targets Type: ")}</b>
                                ${val.target_details || ""}
                        </span><br>
                        <span>
                            <b>${__("Service Type: ")}</b>
                                ${val.disinfection_system}
                        </span><br>
                        <span>
                            <b>${__("Service Address: ")}</b>
                                ${val.customer_address_without_country}
                        </span>
                    </td>
                    <td style="${columns[1].column_width}% !important;">
                        <span>
                            <b>${__("Date: ")}</b>
                            ${val.booking_date}
                        </span><br>
                        <span>
                            <b>${__("Departure: ")}</b>
                            ${val.from_time}
                        </span><br>
                        <span>
                            <b>${__("Start Time: ")}</b>
                            ${val.appointment_start_time}
                        </span><br>
                        <span>
                            <b>${__("ETC: ")}</b>
                                ${val.etc}
                        </span><br><br>
                        <span>
                            <button class='btn btn-primary'>${__("Start")}</button><br><br>
                            <button class="btn btn-info open-support-or-call">${__("Support")}</button>
                        </span>
                    </td>
                    <td style="width:${columns[2].column_width}% !important;vertical-align:middle;" class='text-${frappe.scrub(columns[2].text_align)}'>
                        <b>${val.status}</b>
                    </td>
                <tr>`);
            this.handle_direction($row);
            this.handle_support($row, val);
            this.handle_more_details($row, val);
            return $row
        },
        handle_support: function($row, val){
            var me = this;
            $($row).find(".open-support-or-call").on("click",  (event)=>{
                if(this.support){
                    this.support.handle_existing_support(val);
                    return;
                }
                let args = {
                    "val": val,
                };
                this.support = new germbusters.dashboard.Support(args);
            });
        },
        handle_more_details: function($row, val){
            $row.find(".more-details").on("click", (res)=>{
                if(val.name){
                    frappe.call({
                        "method": "germbusters.www.dashboard.service-technician.index.get_more_details",
                        "args": {
                            "booking_id": val.name,
                        },
                        "freeze": true,
                        "callback": (res)=>{
                            console.log(this);
                        }
                    });
                }
            });
        },
        handle_direction: function($row){
            var me = this;
            $row.find(".btn-primary").on("click", function(event){
                let address = $row.attr("addr");
                let href = frappe.utils.format("https://www.google.com/maps/dir/?api=1&origin={0}&destination={1}",
                                        ["current+location", address]);
                window.open(href, "_blank");
            });
            $row.find(".btn-secondary").on("click", function(event){
                let booking = $row.attr("booking");
                new service_technician.BookingDetail({"booking": booking});
            });
        }
    });

    service_technician.BookingDetail = Class.extend({
        init: function(args){
            $.extend(this, args);
            this.make();
        },
        make: function(){
            this.get_booking_detail();
        },
        get_booking_detail: function(){

            if(!this.booking){
                germbusters.unfreeze();
                return false;
            }
            frappe.unfreeze();
            frappe.unfreeze();
            frappe.call({
                "method": "germbusters.www.dashboard.service-technician.index.get_booking_detail",
                "args":{
                    "booking": this.booking,
                },
                "callback": (res)=>{
                    if(res && res.message){
                        this.init_dialog(res.message)
                    }
                }
            });
        },
        init_dialog: function(detail){

        },
    });
    new service_technician.ServiceTechnician({});

});

File: navdeepghai/dashboard/g_dashboard/book_appointment/book_appointment.py
"""
	Write your python controller here
"""


import frappe
from germbusters.boot import get_default_datetime_format
from germbusters.booking.doctype.book_appointment.book_appointment import pretty_format_minutes
from frappe.utils import getdate, nowdate, get_datetime
from frappe import string_types
import json

def execute(filters={}, dashboard=None):
	return get_booking_list(filters)

def get_booking_list(filters):
	user = frappe.session.user
	booking_date = getdate(filters.get("booking_date"));

	default_format = get_default_datetime_format()
	conditions = get_employee_conditions(user)
	data = []
	results = frappe.db.sql("""SELECT `tabBook Appointment`.name,
	        `tabBook Appointment`.full_name, `tabBook Appointment`.mobile,
	        `tabBook Appointment`.email, `tabBook Appointment`.customer_address,
	        `tabBook Appointment`.to_time_data, `tabBook Appointment`.booking_date,
	        `tabBook Appointment`.customer_address_without_country,
	        `tabBook Appointment`.appointment_start_time,
	        `tabBook Appointment`.from_time, `tabBook Appointment`.to_time,
	        `tabBook Appointment`.target_details, `tabBook Appointment`.disinfection_system,
	        `tabBook Appointment`.total_targets,
	        `tabBook Appointment`.status, `tabBook Appointment Employee`.task
	        FROM `tabBook Appointment` INNER JOIN `tabBook Appointment Employee`
	        ON `tabBook Appointment`.name = `tabBook Appointment Employee`.parent
	        WHERE
	            `tabBook Appointment`.docstatus != 2 AND
	            `tabBook Appointment`.booking_date >= '%s' %s
	        ORDER BY
	            `tabBook Appointment`.appointment_start_time
	        """%(booking_date, conditions), as_dict=True)

	for res in results:
	    time_diff  = ""
	    if(res.appointment_start_time):
	        time_diff = pretty_format_minutes((res.to_time-res.appointment_start_time).seconds)
	    res.update({
	        "booking_date": res.booking_date.strftime(default_format.date_format),
	        "from_time": res.from_time.strftime(default_format.time_format),
	        "appointment_start_time": res.appointment_start_time.strftime(default_format.time_format) if res.appointment_start_time else "",
	        "etc": time_diff
	    })
	    data.append(res)
	return data

# GET SERVICE TECHNICIAN
def get_employee_conditions(user):
	employees = ", ".join(["'%s'"%(e.for_value) for e in frappe.db.sql("""
		SELECT for_value FROM `tabUser Permission` WHERE allow='Employee'
		AND user='%s' """%(user), as_dict=True)])
	conditions = " AND `tabBook Appointment Employee`.employee IN (%s) "%(employees or 'NULL')
	return conditions

@frappe.whitelist()
def get_booking_detail(booking):
    data = {}
    if(booking and frappe.db.exists("Book Appointment", booking)):
        data = frappe.get_doc("Book Appointment", booking).as_dict()
    return data
